File: 2003/VB101SamplesAll/VB101SamplesDataAccess/Creating and using User Defined Types with SQL Server 2005/UDT/GeoCache.vb
Imports System
Imports System.Data
Imports System.Data.Sql
Imports System.Data.SqlTypes
Imports Microsoft.SqlServer.Server

<Serializable()> _
<Microsoft.SqlServer.Server.SqlUserDefinedType(Format.Native)> _
Public Structure GeoCache
    Implements INullable

    ''' <summary>
    ''' This is an overloaded method that allows us to display the data stored
    ''' in the user defined type.
    ''' </summary>
    '' Replace the followng code with your code
    Public Overrides Function ToString() As String

        If Me.m_Null Then
            Return Nothing
        Else
            Return Me.LATITUDE.ToString() & ":" & Me.LONGITUDE.ToString()
        End If
    End Function

    Public ReadOnly Property IsNull() As Boolean Implements INullable.IsNull
        Get
            ' Put your code here
            Return m_Null
        End Get
    End Property

    Public Shared ReadOnly Property Null() As GeoCache
        Get
            Dim h As GeoCache = New GeoCache
            h.m_Null = True
            Return h
        End Get
    End Property

    ''' <summary>
    ''' This property handles access to the private longitude variable
    ''' </summary>
    Public Property LONGITUDE() As Double

        Get
            Return m_longitude
        End Get

        Set(ByVal value As Double)

            m_longitude = value
        End Set

    End Property

    ''' <summary>
    ''' This property handles access ot the private latitude variable
    ''' </summary>
    Public Property LATITUDE() As Double

        Get
            Return m_latitude
        End Get
        Set(ByVal value As Double)
            m_latitude = value

        End Set
    End Property

    Public Shared Function Parse(ByVal s As SqlString) As GeoCache
        If s.IsNull Then
            Return Null
        End If

        Dim u As GeoCache = New GeoCache
        ' Put your code here
        Dim str As String = s.ToString()
        Dim latlong() As String = str.Split(",".ToCharArray())

        u.LATITUDE = Double.Parse(latlong(0))
        u.LONGITUDE = Double.Parse(latlong(1))


        Return u
    End Function


    Private m_Null As Boolean
    Private m_longitude As Double
    Private m_latitude As Double

End Structure

File: 2003/CS101SamplesAll/CS101SamplesDataAccess/Creating and using User Defined Types with SQL Server 2005/UDT/GeoCache.cs
using System;
using System.Data;
using System.Data.Sql;
using System.Data.SqlTypes;
using Microsoft.SqlServer.Server;

[Serializable]
[Microsoft.SqlServer.Server.SqlUserDefinedType(Format.Native)]
public struct GeoCache : INullable
{
	/// <summary>
	/// This is an overloaded method that allows us to display the data stored
	/// in the user defined type.
	/// </summary>
	public override string ToString()
	{
		// Replace the followng code with your code
		if (this.m_Null)
			return null;
		else
			return this.LAT + ":" + this.LONG;
	}
	
	/// <summary>
	/// This method indicates whether the user defined type is null or not
	/// </summary>
	public bool IsNull
	{
		get
		{
			return m_Null;
		}
	}

	/// <summary>
	/// This creates a new instance of the user defined type
	/// </summary>
	public static GeoCache Null
	{
		get
		{
			GeoCache h = new GeoCache();
			h.m_Null = true;
			return h;
		}
	}

	/// <summary>
	/// This property handles access to the private longitude variable
	/// </summary>
	public double LONG
	{
		get
		{
			return m_longitude;
		}

		set
		{
			m_longitude = (double)value;
		}

	}

	/// <summary>
	/// This property handles access ot the private latitude variable
	/// </summary>
	public double LAT
	{
		get
		{
			return m_latitude;
		}
		set
		{
			m_latitude = (double)value;
		}
	}

	/// <summary>
	/// This method parses the values that are sent in and then
	/// sets the user defined type.
	/// </summary>
	public static GeoCache Parse(SqlString s)
	{
		if (s.IsNull)
			return Null;
		else
		{
			GeoCache u = new GeoCache();
			string str = s.ToString();
			string[] latlong = str.Split(",".ToCharArray());

			u.LAT = Double.Parse(latlong[0]);
			u.LONG = Double.Parse(latlong[1]);
			
			return u;
		}
	}

	
	// Private members
	private bool m_Null;
	private double m_longitude;
	private double m_latitude;
	
}


